File: Reserve Your Spot/Services/BookingService.cs
using Reserve_Your_Spot.Models;

namespace Reserve_Your_Spot.Services;

public class BookingService : IBookingService
{
    private readonly SupabaseService _supabase;
    private readonly IBusinessService _businessService;

    public BookingService(SupabaseService supabase, IBusinessService businessService)
    {
        _supabase = supabase;
        _businessService = businessService;
    }

    public async Task<List<TimeSlot>> GetAvailableSlotsAsync(
        string businessId, string? staffId, DateTime date, int durationMinutes)
    {
        try
        {
            // Get business opening hours
            var business = await _businessService.GetBusinessByIdAsync(businessId);
            var dayName = date.DayOfWeek.ToString().ToLower();
            var openTime = TimeSpan.FromHours(9);
            var closeTime = TimeSpan.FromHours(18);

            if (business?.OpeningHours != null && business.OpeningHours.TryGetValue(dayName, out var hours))
            {
                var parts = hours.Split('-');
                if (parts.Length == 2)
                {
                    if (TimeSpan.TryParse(parts[0].Trim(), out var o)) openTime = o;
                    if (TimeSpan.TryParse(parts[1].Trim(), out var c)) closeTime = c;
                }
            }

            // Get existing bookings for that day
            var dateStart = date.Date.ToString("o");
            var dateEnd = date.Date.AddDays(1).ToString("o");
            var filter = staffId != null
                ? $"business_id=eq.{businessId}&staff_id=eq.{staffId}&start_time=gte.{dateStart}&start_time=lt.{dateEnd}&select=*"
                : $"business_id=eq.{businessId}&start_time=gte.{dateStart}&start_time=lt.{dateEnd}&select=*";
            var existingBookings = await _supabase.GetListAsync<Booking>("bookings", filter);

            var slots = new List<TimeSlot>();
            var current = date.Date + openTime;
            var end = date.Date + closeTime;
            var now = DateTime.Now;

            while (current + TimeSpan.FromMinutes(durationMinutes) <= end)
            {
                var slotEnd = current.AddMinutes(durationMinutes);
                var isBooked = existingBookings.Any(b =>
                    b.Status != BookingStatus.Cancelled &&
                    current < b.EndTime && slotEnd > b.StartTime);

                var isPast = current <= now;
                var isLastMinute = !isPast && (current - now).TotalHours <= 2;

                slots.Add(new TimeSlot
                {
                    DateTime = current,
                    IsAvailable = !isBooked && !isPast,
                    StaffId = staffId,
                    IsLastMinute = isLastMinute && !isBooked && !isPast
                });

                current = current.AddMinutes(30); // 30-min intervals
            }

            return slots;
        }
        catch { return new(); }
    }

    public async Task<Booking?> CreateBookingAsync(Booking booking)
    {
        try { return await _supabase.InsertAsync<Booking>("bookings", booking); }
        catch { return null; }
    }

    public async Task<List<Booking>> GetCustomerBookingsAsync(string customerId)
    {
        try
        {
            return await _supabase.GetListAsync<Booking>(
                "bookings", $"customer_id=eq.{customerId}&select=*&order=start_time.desc");
        }
        catch { return new(); }
    }

    public async Task<List<Booking>> GetBusinessBookingsAsync(string businessId, DateTime? date = null)
    {
        try
        {
            var filter = $"business_id=eq.{businessId}&select=*&order=start_time.asc";
            if (date.HasValue)
            {
                var start = date.Value.Date.ToString("o");
                var end = date.Value.Date.AddDays(1).ToString("o");
                filter += $"&start_time=gte.{start}&start_time=lt.{end}";
            }
            return await _supabase.GetListAsync<Booking>("bookings", filter);
        }
        catch { return new(); }
    }

    public async Task<bool> CancelBookingAsync(string bookingId, string reason)
    {
        try
        {
            await _supabase.UpdateAsync("bookings", $"id=eq.{bookingId}",
                new { status = "cancelled", cancellation_reason = reason });
            return true;
        }
        catch { return false; }
    }

    public async Task<bool> ConfirmBookingAsync(string bookingId)
    {
        try
        {
            await _supabase.UpdateAsync("bookings", $"id=eq.{bookingId}", new { status = "confirmed" });
            return true;
        }
        catch { return false; }
    }

    public async Task<bool> MarkNoShowAsync(string bookingId)
    {
        try
        {
            await _supabase.UpdateAsync("bookings", $"id=eq.{bookingId}", new { status = "no_show" });
            return true;
        }
        catch { return false; }
    }

    public async Task<bool> CompleteBookingAsync(string bookingId)
    {
        try
        {
            await _supabase.UpdateAsync("bookings", $"id=eq.{bookingId}", new { status = "completed" });
            return true;
        }
        catch { return false; }
    }
}

File: booklocal-mobile/services/bookingService.ts
import { supabase } from './supabase';
import { Booking, TimeSlot, StaffMember, Business, Service } from '../types';
import { addMinutes, format, parseISO, isBefore, isAfter, setHours, setMinutes } from 'date-fns';

export const bookingService = {
  async getAvailableSlots(
    business: Business,
    service: Service,
    staff: StaffMember | null,
    date: Date
  ): Promise<TimeSlot[]> {
    const dayName = format(date, 'EEE'); // Mon, Tue, etc.
    const slots: TimeSlot[] = [];
    const staffList = staff ? [staff] : (business.staff ?? []);

    for (const s of staffList) {
      const dayHours = s.working_hours?.[dayName];
      if (!dayHours?.is_working) continue;

      const isDayOff = s.days_off?.some(
        (d) => format(parseISO(d), 'yyyy-MM-dd') === format(date, 'yyyy-MM-dd')
      );
      if (isDayOff) continue;

      const [startH, startM] = dayHours.start.split(':').map(Number);
      const [endH, endM] = dayHours.end.split(':').map(Number);

      let current = setMinutes(setHours(new Date(date), startH), startM);
      const endTime = setMinutes(setHours(new Date(date), endH), endM);
      const slotEnd = addMinutes(current, service.duration_minutes);

      // Fetch existing bookings for this staff on this day
      const startOfDay = format(date, "yyyy-MM-dd'T'00:00:00");
      const endOfDay = format(date, "yyyy-MM-dd'T'23:59:59");
      const { data: existingBookings } = await supabase
        .from('bookings')
        .select('start_time, end_time')
        .eq('staff_id', s.id)
        .gte('start_time', startOfDay)
        .lte('start_time', endOfDay)
        .not('status', 'in', '("cancelled","no_show")');

      while (isAfter(endTime, addMinutes(current, service.duration_minutes - 1))) {
        const slotStart = new Date(current);
        const slotEndTime = addMinutes(slotStart, service.duration_minutes);

        const isBooked = (existingBookings ?? []).some((b) => {
          const bStart = parseISO(b.start_time);
          const bEnd = parseISO(b.end_time);
          return slotStart < bEnd && slotEndTime > bStart;
        });

        const isPast = isBefore(slotStart, new Date());

        if (!isPast) {
          slots.push({
            datetime: slotStart.toISOString(),
            is_available: !isBooked,
            staff_id: s.id,
            is_last_minute: false,
          });
        }

        current = addMinutes(current, 30);
      }
    }

    // Deduplicate by time (if "any available" mode picks first free staff)
    const seen = new Set<string>();
    return slots.filter((slot) => {
      const key = slot.datetime;
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    });
  },

  async createBooking(booking: Omit<Booking, 'id' | 'created_at'>): Promise<{ data: Booking | null; error: string | null }> {
    const { data, error } = await supabase
      .from('bookings')
      .insert(booking)
      .select()
      .single();
    return { data: data as Booking | null, error: error?.message ?? null };
  },

  async getCustomerBookings(customerId: string): Promise<Booking[]> {
    const { data } = await supabase
      .from('bookings')
      .select(`
        *,
        business:businesses(id, name, logo_url, address),
        service:services(name, duration_minutes, price),
        staff:staff(name, photo_url)
      `)
      .eq('customer_id', customerId)
      .order('start_time', { ascending: false });
    return (data as Booking[]) ?? [];
  },

  async getBusinessBookings(businessId: string, date?: Date): Promise<Booking[]> {
    let query = supabase
      .from('bookings')
      .select(`
        *,
        customer:users(full_name, phone_number, avatar_url),
        service:services(name, duration_minutes, price),
        staff:staff(name, photo_url)
      `)
      .eq('business_id', businessId)
      .order('start_time');

    if (date) {
      const start = format(date, "yyyy-MM-dd'T'00:00:00");
      const end = format(date, "yyyy-MM-dd'T'23:59:59");
      query = query.gte('start_time', start).lte('start_time', end);
    }

    const { data } = await query;
    return (data as Booking[]) ?? [];
  },

  async cancelBooking(bookingId: string, reason: string): Promise<{ error: string | null }> {
    const { error } = await supabase
      .from('bookings')
      .update({ status: 'cancelled', cancellation_reason: reason })
      .eq('id', bookingId);
    return { error: error?.message ?? null };
  },

  async rescheduleBooking(
    bookingId: string,
    newStartTime: string,
    newEndTime: string
  ): Promise<{ error: string | null }> {
    const { error } = await supabase
      .from('bookings')
      .update({ start_time: newStartTime, end_time: newEndTime, status: 'confirmed' })
      .eq('id', bookingId);
    return { error: error?.message ?? null };
  },

  async confirmBooking(bookingId: string): Promise<{ error: string | null }> {
    const { error } = await supabase
      .from('bookings')
      .update({ status: 'confirmed' })
      .eq('id', bookingId);
    return { error: error?.message ?? null };
  },

  async completeBooking(bookingId: string): Promise<{ error: string | null }> {
    const { error } = await supabase
      .from('bookings')
      .update({ status: 'completed' })
      .eq('id', bookingId);
    return { error: error?.message ?? null };
  },

  async markNoShow(bookingId: string): Promise<{ error: string | null }> {
    const { error } = await supabase
      .from('bookings')
      .update({ status: 'no_show' })
      .eq('id', bookingId);
    return { error: error?.message ?? null };
  },

  async createManualBooking(booking: Partial<Booking>): Promise<{ data: Booking | null; error: string | null }> {
    const payload = { ...booking, is_manual: true, status: 'confirmed', deposit_paid: false };
    const { data, error } = await supabase.from('bookings').insert(payload).select().single();
    return { data: data as Booking | null, error: error?.message ?? null };
  },

  async getBusinessAnalytics(businessId: string, startDate: Date, endDate: Date): Promise<{
    total_bookings: number;
    total_revenue: number;
    new_customers: number;
    no_shows: number;
    cancellations: number;
    deposit_collected: number;
  }> {
    const { data } = await supabase
      .from('bookings')
      .select('status, total_price, deposit_amount, deposit_paid, customer_id')
      .eq('business_id', businessId)
      .gte('start_time', startDate.toISOString())
      .lte('start_time', endDate.toISOString());

    const bookings = data ?? [];
    const uniqueCustomers = new Set(bookings.map((b: any) => b.customer_id)).size;

    return {
      total_bookings: bookings.length,
      total_revenue: bookings
        .filter((b: any) => b.status === 'completed')
        .reduce((sum: number, b: any) => sum + (b.total_price ?? 0), 0),
      new_customers: uniqueCustomers,
      no_shows: bookings.filter((b: any) => b.status === 'no_show').length,
      cancellations: bookings.filter((b: any) => b.status === 'cancelled').length,
      deposit_collected: bookings
        .filter((b: any) => b.deposit_paid)
        .reduce((sum: number, b: any) => sum + (b.deposit_amount ?? 0), 0),
    };
  },
};
